Feature / Hydration AMM Connector and Polkadot Chain

Danilo-Araujo-Silva opened the pull request:
Adding Polkadot chain and Hydration AMM connector.

This implementation is linked to the following NCP: [NCP-23: Hydration AMM Connector and Polkadot Chain](https://snapshot.box/#/s:hbot-ncp.eth/proposal/0x701eed6edab58f8c43d800bd1486df3d39cd3d1a89605d1ea3c3f1a0b56166e9)

Hummingbot Gateway core 2.6 compatible PR: https://github.com/hummingbot/gateway/pull/447
Hummingbot Client PR: https://github.com/hummingbot/hummingbot/pull/7572
Hummingbot Site documentation PR: https://github.com/hummingbot/hummingbot-site/pull/442

Some useful resources for testing:
- We recommend using [Talisman Wallet](https://talisman.xyz/) and have funds in the Hydration parachain, which can be checked [here](https://app.hydration.net/wallet/assets) (especially in HDX, DOT, USDT, and USDC (prefer Polkadot, not the Wormhole one, or edit the src/template/lists/hydration.json) tokens)
- To transfer funds from another chain to Hydration parachain their [cross exchange functionality](https://app.hydration.net/cross-chain) can be used
- The main Hydration site is available [here](https://hydration.net/), their web app [here](https://app.hydration.net/trade/swap), and some documentation links are [here](https://docs.hydration.net/), [here](https://apidocs.bsx.fi/Hydration), and [here](https://github.com/galacticcouncil/sdk)
- For testing the endpoints one can use [this .http file](https://github.com/robotter-ai/gateway/blob/fix/pr/feat/hydration/temporary/requests/http/hydration.http) we have used

Danilo-Araujo-Silva reviewed (commented):
All changes were reviewed, we tried to touch the core files as less as possible. Please note that the `connection-manager` file needed more intervention because the ChainInstance was in need to be used.

Danilo-Araujo-Silva commented:
@nikspz, I have updated the PR description to help the Q&A team with the testing. Please let us know if you have any questions.

I may also periodically push force to avoid merge commits and try to keep the PR as just one single commit.

nikspz commented:
Not showing balance, idk if related to the default node

![image](https://github.com/user-attachments/assets/1f1e68df-4aaa-4d81-97e8-e952a5d93890)

Danilo-Araujo-Silva commented:
> Isn't it using recovery phrase (not private key)? @Danilo-Araujo-Silva ![image](https://private-user-images.githubusercontent.com/83953535/437067830-21e7f8d0-6870-421f-b164-a5f72ec9fe48.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1MTU0NDEsIm5iZiI6MTc0NTUxNTE0MSwicGF0aCI6Ii84Mzk1MzUzNS80MzcwNjc4MzAtMjFlN2Y4ZDAtNjg3MC00MjFmLWIxNjQtYTVmNzJlYzlmZTQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI0VDE3MTkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0OTlkYTgwMjAwOGIzY2FjNTYxZDMzZjAyNDNhNmJhNjQ5ZDQwYTg0NjU4NDNlYTFhYjk5YjRmYjg4NDRmOGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hn5mBMi8HHB8wj5qcNT0PACbo1vXmMA9bDZ53NBn4vo)

@nikspz , yes, indeed, for Polkadot it is needed to inform the wallet mnemonic instead of a private key.

Unfortunately, we cannot change this message because it is automatically formed on HB client [here](https://github.com/hummingbot/hummingbot/blob/487999b53d67cfb0d829e5a42ec993e840268e25/hummingbot/client/command/gateway_command.py#L388) in a generic, non customized, way.

Danilo-Araujo-Silva commented:
> Not showing balance, idk if related to the default node
> 
> ![image](https://private-user-images.githubusercontent.com/83953535/437102663-1f1e68df-4aaa-4d81-97e8-e952a5d93890.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1MTU0NDEsIm5iZiI6MTc0NTUxNTE0MSwicGF0aCI6Ii84Mzk1MzUzNS80MzcxMDI2NjMtMWYxZTY4ZGYtNGFhYS00ZDgxLTk3ZTgtZTk1MmE1ZDkzODkwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI0VDE3MTkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwNjhjYjdkNGIyMjNiZjdkNDUwMDUxNTU3ODI0NTFmZThhYzExM2RlYmZkNmEzZDA2Nzk5MjRkNWEzYzkzNmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vLsNl8hH9vfuKfB78496QKjjX7R93upcFEfsKLty4Mg)

About this error, the problem is the current version of the HB client has a hardcoded information about the supported chains. Because we are still working on a script for HB Client, please kindly change this [line](https://github.com/hummingbot/hummingbot/blob/487999b53d67cfb0d829e5a42ec993e840268e25/hummingbot/client/settings.py#L108) adding polkadot there:

Before:

> for chain in ["ethereum", "solana"]:

After:

> for chain in ["ethereum", "solana", "polkadot"]:

After that the balance should work:

> gateway connector-tokens hydration/amm_polkadot_mainnet HDX,DOT,USDC,USDT,SOL
> gateway balance

Regarding the `Abnormal closure error`, this indeed is happening with their RPC, but we have greatly mitigated the impact of it instantiating objects again and using an automatic retry mechanism. So please consider it as an ignorable error for now.

If you still happen to find inconsistencies, please kindly let us know.

PS.: We have seen that the `hummingbot/core/gateway/gateway_http_client.py` implementation is not fully supporting for AMM connectors, but focusing more on the CLMM ones, but some basic functionality should be ok. We will be preparing an HB Client PR soon (which will need to change this core file). If you guys find inconsistencies when using the HB Client in its current state, we can speed it up. In this case, we would like to kindly ask you guys to test this PR of ours outside the HB Client, calling the HB Gateway endpoints directly.

Danilo-Araujo-Silva commented:
@nikspz,

Here is our draft PR for the HB Client (still in draft mode), but as you will be able to see, it basically changes the settings.py, the gateway_http_client.py and include new scripts.

https://github.com/hummingbot/hummingbot/pull/7572

Danilo-Araujo-Silva commented:
@nikspz, when you have time, please let us know if the tests are successful. Thanks in advance.

Danilo-Araujo-Silva commented:
> * commit [d50e96b](https://github.com/hummingbot/gateway/commit/d50e96b6a74808641e466d70bf4242ad80887b3c) + commit e4c97aa
>   
>   * test performed:
>     
>     * cloned and installed PR7572 + gwPR441
>     * connected hydration successfully
>     * checked balance successfully for hydration/amm_polkadot_mainnet
>     * connector-tokens: ok, added token, balance shows: ok
>       amm_arb: I think I got Getting rate limited by default node
> 
> [logs_gateway_app.log.2025-04-29.log](https://github.com/user-attachments/files/19965408/logs_gateway_app.log.2025-04-29.log) [logs_conf_amm_arb_2.log](https://github.com/user-attachments/files/19965419/logs_conf_amm_arb_2.log) [logs_conf_amm_arb_1.log](https://github.com/user-attachments/files/19965421/logs_conf_amm_arb_1.log)
> 
> ![image](https://private-user-images.githubusercontent.com/83953535/438916914-3ce8ab3d-c2bc-444e-a190-332be37fb4f9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU5NjQyNTMsIm5iZiI6MTc0NTk2Mzk1MywicGF0aCI6Ii84Mzk1MzUzNS80Mzg5MTY5MTQtM2NlOGFiM2QtYzJiYy00NDRlLWExOTAtMzMyYmUzN2ZiNGY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI5VDIxNTkxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY4ZjYwYzUxODk3NzQ1YmM0NTZhNWVhMDdlZDRlNzBiNWY4ZWRlMTJhZjUxNGRkZjE1M2ZiMjA2NGFhOTQwYzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OCrCIkAgl8EQDRxlPwWdxQbSCGJSOQKdNPUroMnclNk)

@nikspz, using your tests we have found a bug, sometimes the wallet was not being decrypted correctly and this was the cause of the error you pointed. Because we have also found (on other files) room for some improvements, we have also done that and pushed.

I'm not sure if the current `amm_arb` strategy is fully compatible with the new AMM connectors (which is the case for Hydration and Raydium, for example). But I was able to see that the swaps are working there (where, depending on the token pair used, some temporary errors could happen). In this case, please let us know if you find any persistent inconsistencies.

We have also tested the following v2 scripts (which were successful, please note that in the `amm_data_feed_example` you need to change the code directly):
- scripts/amm_price_example.py
- scripts/amm_trade_example.py
- scripts/amm_data_feed_example.py

When you have time, please take a look for us again.

Many thanks in advance! o/

Danilo-Araujo-Silva commented:
> @Danilo-Araujo-Silva Those PRs PR7572 + gwPR441 will need to be reviewed after 2.6.0 (client and gateway release) since major changes coming in the next release. for [hummingbot/hummingbot#7572](https://github.com/hummingbot/hummingbot/pull/7572) this affected files Hydration PR scope, will be dev reviewed later. https://github.com/hummingbot/hummingbot/pull/7572/files#diff-4b217ba56fd93bbbefbcc4e90c942f81a35200b05b47a1e73571d9989682916eL318
> 
> commit [0733531](https://github.com/hummingbot/gateway/commit/0733531fc538712790b26f5fafb98d0a9656f017) + commit a9285fc
> 
> * cloned and installed PR7572 + gwPR441
> * connected hydration successfully
> * checked balance successfully for hydration/amm_polkadot_mainnet
>   
>   * when check balance have warnings
>     
>     * REGISTRY: Unknown signed extensions ValidateClaim found, treating them as no-effect
>     * API/INIT: RPC methods not decorated: chainHead_v1_body, chainHead_v1_call, chainHead_v1_continue, chainHead_v1_follow, chainHead_v1_header, chainHead_v1_stopOperation, chainHead_v1_storage, chainHead_v1_unfollow, chainHead_v1_unpin, eth_getEthGas, eth_getEthResources, eth_isBlockFinalized, eth_isTransactionFinalized, eth_sign, net_health, net_indexer, net_isSafeMode, net_runtimeVersion
> * connector-tokens: ok, added token, balance shows: ok
> * scripts/amm_price_example.py: showing price successfully, after stop command still checking price and have some node related warnings: API-WS: disconnected from wss://rpc.hydradx.cloud: 1006:: Abnormal Closure
> * scripts/amm_trade_example.py
>   
>   * Created LIMIT SELL order
>   * initial trade executed successfully https://hydration.subscan.io/extrinsic/0x7f24848a14c6115bf0f5b950e5f4a1c9a7b090cc5c8d35033d0682974ad8743d
>   * then bot stuck with NotImplementedError: Unsupported chain: polkadot
> 
> [logs_hdx-sell.log](https://github.com/user-attachments/files/20065169/logs_hdx-sell.log)
> 
> amm_price_example.py [logs_price-sell1.log](https://github.com/user-attachments/files/20065282/logs_price-sell1.log)
> 
> ![image](https://private-user-images.githubusercontent.com/83953535/440840789-26e1091e-aeb6-4c2f-9c14-478d3816b933.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2NDc0NjgsIm5iZiI6MTc0NjY0NzE2OCwicGF0aCI6Ii84Mzk1MzUzNS80NDA4NDA3ODktMjZlMTA5MWUtYWViNi00YzJmLTljMTQtNDc4ZDM4MTZiOTMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA3VDE5NDYwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmYmI1NmE1NTM2YzhhZmM2YTk0NzgxM2U4M2MyYmEyN2NmY2I1MGU2YTVmOGIxMmQ0NzkyNzI3NDAzNjUxZWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.l-Jy0kng9FswEBvFJJRP2o1pZtkCWvuqAwceefvf6k0)
> 
> amm_trade_example.py ![hdx-sell](https://private-user-images.githubusercontent.com/83953535/440840915-f0dfedff-10da-49de-bb78-15d53493343f.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2NDc0NjgsIm5iZiI6MTc0NjY0NzE2OCwicGF0aCI6Ii84Mzk1MzUzNS80NDA4NDA5MTUtZjBkZmVkZmYtMTBkYS00OWRlLWJiNzgtMTVkNTM0OTMzNDNmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA3VDE5NDYwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4Yzk2NzZmMWI4YzBiZmNjNmIwNTNiMzlhMGE2M2ZkOGJhYjdkMTI1NzBjNjIwNDdlZWRjODE5YTZmZDgzMTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.P8AspKS4f6rsBsxmch8CZXmNXlFwY-lZRfGqXwDBYbA) [ ![hdx-sell](https://private-user-images.githubusercontent.com/83953535/440840915-f0dfedff-10da-49de-bb78-15d53493343f.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2NDc0NjgsIm5iZiI6MTc0NjY0NzE2OCwicGF0aCI6Ii84Mzk1MzUzNS80NDA4NDA5MTUtZjBkZmVkZmYtMTBkYS00OWRlLWJiNzgtMTVkNTM0OTMzNDNmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA3VDE5NDYwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4Yzk2NzZmMWI4YzBiZmNjNmIwNTNiMzlhMGE2M2ZkOGJhYjdkMTI1NzBjNjIwNDdlZWRjODE5YTZmZDgzMTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.P8AspKS4f6rsBsxmch8CZXmNXlFwY-lZRfGqXwDBYbA) ](https://private-user-images.githubusercontent.com/83953535/440840915-f0dfedff-10da-49de-bb78-15d53493343f.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2NDc0NjgsIm5iZiI6MTc0NjY0NzE2OCwicGF0aCI6Ii84Mzk1MzUzNS80NDA4NDA5MTUtZjBkZmVkZmYtMTBkYS00OWRlLWJiNzgtMTVkNTM0OTMzNDNmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA3VDE5NDYwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4Yzk2NzZmMWI4YzBiZmNjNmIwNTNiMzlhMGE2M2ZkOGJhYjdkMTI1NzBjNjIwNDdlZWRjODE5YTZmZDgzMTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.P8AspKS4f6rsBsxmch8CZXmNXlFwY-lZRfGqXwDBYbA) [ ](https://private-user-images.githubusercontent.com/83953535/440840915-f0dfedff-10da-49de-bb78-15d53493343f.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2NDc0NjgsIm5iZiI6MTc0NjY0NzE2OCwicGF0aCI6Ii84Mzk1MzUzNS80NDA4NDA5MTUtZjBkZmVkZmYtMTBkYS00OWRlLWJiNzgtMTVkNTM0OTMzNDNmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA3VDE5NDYwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4Yzk2NzZmMWI4YzBiZmNjNmIwNTNiMzlhMGE2M2ZkOGJhYjdkMTI1NzBjNjIwNDdlZWRjODE5YTZmZDgzMTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.P8AspKS4f6rsBsxmch8CZXmNXlFwY-lZRfGqXwDBYbA)

Hi @nikspz, thanks for getting back to us

Regarding the merging, it's really important for us to have the code merged altogether with the next release. That's why we have conformed to the new architecture, and we are up to date with the dev branch. If needed, we can be in sync with the `core-2.6-mcp` branch (which seems to be the most updated one). Actually I have opened this draft PR here which I can make it our official, if needed, this draft PR is ready to be merged.

Please take a look at it here:
https://github.com/hummingbot/gateway/pull/445

About the HB Client, it seems the dev branch is still the most recent one and our PR seems to be compatible with it after I rebased it. Please take this into consideration, the needed changes is basically the ones in the gateway_http_client.py (to have the amm endpoints) and the settings.py (to enable the polkadot chain to work too). (Maybe `Unsupported chain: polkadot` error might point to another modification needed on the Client side, we will take a look here, but it's a bit strange since we have tested all these amm scripts).

About the tests, thanks for getting back to us on this regard too.
Regarding the warnings and the WSS abornal closure and the other warnings (we will try to mitigate them but they should not be creating problems with some code we had written, in this case, please desconsider them for the merging for now, and we will be addressing them with some new features we would like to add in the future (they happen independently inside of their library and they are more difficult to handle, maybe some modifications on their library is necessary)).

About the amm_price continuous running, this is more related with the HB Client architecture and the script implementation (and since this script is not ours, please desconsider it as a blocker for us to get merged, but indeed if the price was not coming this would have been a problem)

Danilo-Araujo-Silva commented:
@nikspz, about the `Unsupported chain: polkadot`, we have seen this is because of new changes that came to the HB Client after I rebased the branch that changed the support for the chains there (I'd say that would be better to standardize this on  the gateway so no change on the client should be needed). In any case, we will be preparing a modification soon and the HB Client amm strategies should be fully working again.

nikspz commented:
Thanks @Danilo-Araujo-Silva, we need to merge https://github.com/hummingbot/gateway/pull/444 first

fengtality commented:
https://github.com/hummingbot/gateway/pull/444 is still in flux and won't be done until staging for v2.6 in approximately 2 weeks.

After the v2.6 release, Gateway should have the Ethereum/Uniswap connectors along with the Solana DEX connectors, plus comprehensive unit tests, and documentation on how new chains and connectors can be added to the new standard. Until then, our team will not review any new connectors.

MHHukiewitz commented:
> https://github.com/hummingbot/gateway/pull/444 is still in flux and won't be done until staging for v2.6 in approximately 2 weeks.
> 
> 
> 
> After the v2.6 release, Gateway should have the Ethereum/Uniswap connectors along with the Solana DEX connectors, plus comprehensive unit tests, and documentation on how new chains and connectors can be added to the new standard. Until then, our team will not review any new connectors.

So we got #447 which would be compatible with #444, but we can't merge this PR because it wouldn't be compatible with #444? Wasn't this PR already thoroughly reviewed by @nikspz?

fengtality commented:
> So we got #447 which would be compatible with #444, but we can't merge this PR because it wouldn't be compatible with #444? Wasn't this PR already thoroughly reviewed by @nikspz?

#444 is still in flux because it is defining the new standard. It doesn't make sense to conduct engineering or QA review for any new connectors until we have the final standards and the testing guidelines for them.
